docs/source/jit.rst

```diff
@@ -531,13 +531,253 @@ Python-defined Constants
+                  %14 : Dynamic = aten::select(%0, %12, %13)
+                  %15 : Dynamic = aten::mul(%11, %14)
+              +   %16 : int = prim::Constant[value=0]()
+              +   %17 : int = prim::Constant[value=3]()
+              +   %18 : Dynamic = aten::select(%0, %16, %17)
+              +   %19 : Dynamic = aten::mul(%15, %18)
+              -   return (%15);
+              ?             ^
+              +   return (%19);
+              ?             ^
+            }
+
+
+    This message indicates to us that the computation differed between when
+    we first traced it and when we traced it with the ``check_inputs``. Indeed,
+    the loop within the body of ``loop_in_traced_fn`` depends on the shape
+    of the input ``x``, and thus when we try another ``x`` with a different
+    shape, the trace differs.
+
+    In this case, data-dependent control flow like this can be captured using
+    script instead::
+
+        def fn(x):
+            result = x[0]
+            for i in range(x.size(0)):
+                result = result * x[i]
+            return result
+
+        inputs = (torch.rand(3, 4, 5),)
+        check_inputs = [(torch.rand(4, 5, 6),), (torch.rand(2, 3, 4),)]
+
+        scripted_fn = torch.jit.script(fn)
+        print(scripted_fn.graph)
+
+        for input_tuple in [inputs] + check_inputs:
+            torch.testing.assert_allclose(fn(*input_tuple), scripted_fn(*input_tuple))
+
+
+    Which produces::
+
+        graph(%x : Dynamic) {
+          %1 : int = prim::Constant[value=0]()
+          %2 : int = prim::Constant[value=0]()
+          %result.1 : Dynamic = aten::select(%x, %2, %1)
+          %4 : int = aten::size(%x, %1)
+          %5 : int = prim::Constant[value=1]()
+          %result : Dynamic = prim::Loop(%4, %5, %result.1)
+            block0(%i : int, %7 : Dynamic) {
+              %9 : int = prim::Constant[value=0]()
+              %10 : Dynamic = aten::select(%x, %9, %i)
+              %result.2 : Dynamic = aten::mul(%7, %10)
+              %12 : int = prim::Constant[value=1]()
+              -> (%12, %result.2)
+            }
+          return (%result);
+        }
+
+
+Tracer Warnings
+    The tracer produces warnings for several problematic patterns in traced
+    computation. As an example, take a trace of a function that contains an
+    in-place assignment on a slice (a view) of a Tensor::
+
+        def fill_row_zero(x):
+            x[0] = torch.rand(*x.shape[1:2])
+            return x
+
+        traced = torch.jit.trace(fill_row_zero, (torch.rand(3, 4),))
+        print(traced.graph)
+
+
+    Produces several warnings and a graph which simply returns the input::
+
+        update_view.py:4: TracerWarning: There are 2 live references to the tensor being modified when tracing in-place operator copy_ (possibly due to an assignment) which  might cause the trace to be incorrect. We can't record the data flow of  Python values, which means the trace might not generalize to other inputs.
```

Comment:
nit: Adam has a PR out that changes the text of these warnings.

Reply:
The warning PR has already landed.